arparser/src/parser.sk

```diff
@@ -0,0 +1,614 @@
+    if (this.scanned) {
+      return Success(void)
+    };
+    this.parse_global_header() match {
+    | Failure(err) -> return Failure(err)
+    | Success _ -> void
+    };
+    loop {
+      header_start = this.next_entry_start;
+
+      this.reader.seek(IO.SeekFromStart(header_start)) match {
+      | Failure(err) -> return Failure(IOError(err))
+      | Success _ -> void
+      };
+
+      header = Header::read(this.reader) match {
+      | Failure(err) -> return Failure(err)
+      | Success(None()) -> break void
+      | Success(Some(v)) -> v
+      };
+
+      header_len = kENTRY_HEADER_LEN;
+      if (
+        !(this.variant is VariantBSD _) &&
+        header.identifier.startsWith("/")
+      ) {
+        this.!variant = VariantGNU();
+        if (header.identifier == kGNU_NAME_TABLE_ID) {
+          this.name_table.resize(header.size, UInt8::truncate(0));
+          this.reader.read_exact(this.name_table.mbytes()) match {
+          | Failure(err) -> return Failure(IOError(err))
+          | Success _ -> void
+          }
+        } else if (header.identifier != kGNU_SYMBOL_LOOKUP_TABLE_ID) {
+          !header.identifier = this.parse_gnu_extended_identifier(
+            header,
+          ) match {
+          | Failure(err) -> return Failure(err)
+          | Success(v) -> v
+          }
+        }
+      } else if (
+        !(this.variant is VariantBSD _) &&
+        header.identifier.endsWith("/")
+      ) {
+        this.!variant = VariantGNU();
+        !header.identifier = header.identifier.stripSuffix("/")
+      } else if (
+        !(this.variant is VariantGNU _) &&
+        header.identifier.startsWith(kBSD_EXTENDED_IDENTIFIER_PREFIX)
+      ) {
+        this.!variant = VariantBSD();
+        (
+          extended_identifier,
+          header_extra_len,
+        ) = this.parse_bsd_extended_identifier(header) match {
+        | Failure(err) -> return Failure(err)
+        | Success(v) -> v
+        };
+        !header.identifier = extended_identifier;
+        !header.size = header.size - header_extra_len;
+        !header_len = header_len + header_extra_len
+      };
+
+      this.!next_entry_start =
+        this.next_entry_start +
+        header_len +
+        header.size +
+        (header.size % 2);
+
+      if (this.is_name_table_id(header.identifier)) {
+        continue
+      };
+
+      entry = HeaderAndLocation{
+        header,
+        header_start,
+        data_start => header_start + header_len,
+      };
+
+      if (this.is_symbol_lookup_table_id(header.identifier)) {
+        this.!symbol_table_header = Some(entry)
+      } else {
+        this.headers.push(entry)
+      }
+    };
+
+    this.!scanned = true;
+    Success(void)
+  }
+
+  private mutable fun parse_gnu_extended_identifier(
+    header: Header,
+  ): Result<String, Error> {
+    start = parse_number(
+      header.identifier
+        .bytes()
+        .slice(kENTRY_IDENTIFIER_OFFSET + 1, kENTRY_IDENTIFIER_LEN)
```

Comment:
`kENTRY_IDENTIFIER_OFFSET + 1` -> `1` because you're in the bytes of `identifier`, not `header`